remoting/kubernetes/client.go

```diff
@@ -0,0 +1,655 @@
+	}
+
+	currentPodName, err := getCurrentPodName()
+	if err != nil {
+		return nil, perrors.WithMessage(err, "get pod name")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c := &Client{
+		currentPodName: currentPodName,
+		ns:             namespace,
+		cfg:            cfg,
+		rawClient:      rawClient,
+		ctx:            ctx,
+		store:          newStore(ctx),
+		cancel:         cancel,
+	}
+
+	currentPod, err := c.initCurrentPod()
+	if err != nil {
+		return nil, perrors.WithMessage(err, "init current pod")
+	}
+
+	// record current status
+	c.currentPod = currentPod
+
+	// init the store by current pods
+	if err := c.initStore(); err != nil {
+		return nil, perrors.WithMessage(err, "init store")
+	}
+
+	// start kubernetes watch loop
+	if err := c.maintenanceStatus(); err != nil {
+		return nil, perrors.WithMessage(err, "maintenance the kubernetes status")
+	}
+
+	logger.Info("init kubernetes registry success")
+	return c, nil
+}
+
+// initCurrentPod
+// 1. get current pod
+// 2. give the dubbo-label for this pod
+func (c *Client) initCurrentPod() (*v1.Pod, error) {
+
+	// read the current pod status
+	currentPod, err := c.rawClient.CoreV1().Pods(c.ns).Get(c.currentPodName, metav1.GetOptions{})
+	if err != nil {
+		return nil, perrors.WithMessagef(err, "get current (%s) pod in namespace (%s)", c.currentPodName, c.ns)
+	}
+
+	oldPod, newPod, err := c.assembleDUBBOLabel(currentPod)
+	if err != nil {
+		if err != ErrDubboLabelAlreadyExist {
+			return nil, perrors.WithMessage(err, "assemble dubbo label")
+		}
+		// current pod don't have label
+	}
+
+	p, err := c.getPatch(oldPod, newPod)
+	if err != nil {
+		return nil, perrors.WithMessage(err, "get patch")
+	}
+
+	currentPod, err = c.patchCurrentPod(p)
+	if err != nil {
+		return nil, perrors.WithMessage(err, "patch to current pod")
+	}
+
+	return currentPod, nil
+}
+
+// initStore
+// 1. get all with dubbo label pods
+// 2. put every element to store
+func (c *Client) initStore() error {
+
+	pods, err := c.rawClient.CoreV1().Pods(c.ns).List(metav1.ListOptions{
+		LabelSelector: fields.OneTermEqualSelector(DubboIOLabelKey, DubboIOLabelValue).String(),
+	})
+	if err != nil {
+		return perrors.WithMessagef(err, "list pods  in namespace (%s)", c.ns)
+	}
+
+	for _, pod := range pods.Items {
+		logger.Debugf("got the pod (name: %s), (label: %v), (annotations: %v)", pod.Name, pod.GetLabels(), pod.GetAnnotations())
+		c.handleWatchedPodEvent(&pod, watch.Added)
+	}
+
+	return nil
+}
+
+// maintenanceStatus
+// try to watch kubernetes pods
+func (c *Client) maintenanceStatus() error {
+
+	c.wg.Add(1)
+
+	// try once
+	watcher, err := c.rawClient.CoreV1().Pods(c.ns).Watch(metav1.ListOptions{
+		LabelSelector: fields.OneTermEqualSelector(DubboIOLabelKey, DubboIOLabelValue).String(),
+		Watch:         true,
+	})
+	if err != nil {
+		return perrors.WithMessagef(err, "try to watch the namespace (%s) pods", c.ns)
```

Comment:
If you return here, c.wg.Add(1) but you do not start a goroutine. Is this a problem?

Reply:
Yes, this is a bad logic. I had fix this in latest commit.